File: services/productService.js
const ProductSQL = require("../models/product.sql");
const ProductMongo = require("../models/product.mongo");

// -------------------- GET ALL --------------------
const getAllProducts = async () => {
  const sqlData = await ProductSQL.findAll();
  const mongoData = await ProductMongo.find();

  return {
    sql: sqlData,
    mongo: mongoData
  };
};

// -------------------- GET BY ID --------------------
const getProductById = async (id) => {
  const sqlData = await ProductSQL.findByPk(id);

  // Mongo uses _id, so this may differ
  let mongoData = null;
  if (id.length === 24) {
    mongoData = await ProductMongo.findById(id);
  }

  return {
    sql: sqlData,
    mongo: mongoData
  };
};

// -------------------- CREATE --------------------
const createProduct = async (data) => {
  const sqlData = await ProductSQL.create({
    name: data.name,
    price: data.price,
    category: data.category,
    stock: data.stock
  });

  const mongoData = await ProductMongo.create(data);

  return {
    sql: sqlData,
    mongo: mongoData
  };
};

// -------------------- UPDATE --------------------
const updateProduct = async (id, data) => {
  let sqlData = null;
  let mongoData = null;

  // SQL
  const productSQL = await ProductSQL.findByPk(id);
  if (productSQL) {
    await productSQL.update(data);
    sqlData = productSQL;
  }

  // Mongo (only if valid ObjectId)
  if (id.length === 24) {
    mongoData = await ProductMongo.findByIdAndUpdate(id, data, {
      new: true
    });
  }

  return { sql: sqlData, mongo: mongoData };
};

// -------------------- DELETE --------------------
const deleteProduct = async (id) => {
  let sqlData = null;
  let mongoData = null;

  // SQL
  const productSQL = await ProductSQL.findByPk(id);
  if (productSQL) {
    await productSQL.destroy();
    sqlData = productSQL;
  }

  // Mongo
  if (id.length === 24) {
    mongoData = await ProductMongo.findByIdAndDelete(id);
  }

  return { sql: sqlData, mongo: mongoData };
};

module.exports = {
  getAllProducts,
  getProductById,
  createProduct,
  updateProduct,
  deleteProduct
};

File: Fast_API/E-commerce/app/services/product_service.py
products = [
  {
    "id" : 1,
    "name" : "laptop",
    "price" : 50000,
    "category" : "Electronics",
    "stock" : 10
  },
  {
    "id" : 2,
    "name" : "mouse",
    "price" : 500,
    "category" : "Electronics",
    "stock" : 50
  }
  
]

def get_all_products():
  return products

  # GET Products By Id

def get_product_by_id(product_id: int):
  for product in products:
    if product["id"] == product_id:
      return product
  return None

def create_product(product_data):
  new_product = product_data.dict()
  new_product["id"] = len(products) + 1
  products.append(new_product)
  return new_product

def update_product(product_id: int, product_data):
  for index, product in enumerate(products):
    if product["id"] == product_id:
      products[index].update(product_data.dict(exclude_unset=True))
      return products[index]
  return None

def delete_product(product_id: int):
  for index, product in enumerate(products):
    if product["id"] == product_id:
      return products.pop(index)
  return None

def replace_product(product_id: int, product_data):
  for index, product in enumerate(products):
    if product["id"] == product_id:
      replaced = product_data.dict()
      replaced["id"] = product_id
      products[index] = replaced
      return products[index]
  return None